### PR TITLE
ELPP-3284 Annotation listing

### PR DESCRIPTION
**Pull request by davidcmoulton**

### CODE REVIEW

File: src/ViewModel/AnnotationTeaser.php [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+use eLife\Patterns\ViewModel;
+use Traversable;
+
+final class AnnotationTeaser implements ViewModel
+{
+    use ArrayAccessFromProperties;
+    use ArrayFromProperties;
+    use SimplifyAssets;
+
+    const RESTRICTED_ACCESS = true;
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: What's this for?

File: src/ViewModel/AnnotationTeaser.php [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        if ($isRestrictedAccess) {
+            $this->meta = Meta::withText(self::RESTRICTED_ACCESS_TEXT, $date);
+        } else {
+            $this->meta = Meta::withDate($date);
+        }
+    }
+
+    public static function full(
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: Is there a better name for this? All the things probably shouldn't be optional too?

File: src/ViewModel/AnnotationTeaser.php (hunk 1 above)
Comment (davidcmoulton) [reply src/ViewModel/AnnotationTeaser.php]: This is for a private annotation that is only visible to the logged in user who made it.

Comment (thewilkybarkid) [reply src/ViewModel/AnnotationTeaser.php]: Ah, see it in the tests now. A bit verbose to wrap a boolean.

File: src/ViewModel/AnnotationTeaser.php (hunk 2 above)
Comment (thewilkybarkid) [reply src/ViewModel/AnnotationTeaser.php]: Looking again, `$inContextUri`, `$highlight` and `$content` can't be empty.

File: src/ViewModel/AnnotationTeaser.php [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        string $document,
+        Date $date = null,
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: Not allowed to be `null`?

File: src/ViewModel/AnnotationTeaser.php [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            $inContextUri,
+            null,
+            $content,
+            $isRestricted
+        );
+    }
+
+    public static function forReply(
+        string $document,
+        Date $date = null,
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: Not allowed to be `null`?

File: src/ViewModel/AnnotationTeaser.php [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        string $document,
+        Date $date,
+        string $inContextUri,
+        string $highlight = null,
+        string $content = null,
+        bool $isRestrictedAccess = false,
+        bool $isReply = false
+    ) {
+        Assertion::notEmpty($document);
+        Assertion::notNull($date);
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: The type hint doesn't allow it to be `null`.

File: src/ViewModel/AnnotationTeaser.php [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    private function __construct(
+        string $document,
+        Date $date,
+        string $inContextUri,
+        string $highlight = null,
+        string $content = null,
+        bool $isRestrictedAccess = false,
+        bool $isReply = false
+    ) {
+        Assertion::notEmpty($document);
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: Prefer `notBlank()` for all these.

File: src/ViewModel/AnnotationTeaser.php [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        Assertion::notEmpty($inContextUri);
+
+        $this->document = $document;
+        $this->inContextUri = $inContextUri;
+
+        if ($isReply) {
+            $this->isReply = $isReply;
+        }
+
+        if ($highlight) {
```
Comment (thewilkybarkid) [inline src/ViewModel/AnnotationTeaser.php]: `if` when it's a `string`|`null` isn't needed.

**Pull request merged**